### PR TITLE
Tim Chow (basics-11-14)

### PR DESCRIPTION
**Pull request by timochow**
Please fill out the survey before submitting the pull request. Thanks!

🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀🚀

How many hours did you spend on this assignment? 
3

Please fill in one error and/or error message you received while working on this assignment.
- In tracking the number of wins, it was not feasible to initialise an array to a fixed length, based on the no. of players. Instead, I populated the array as I went along. This led to an issue where some values in the array were 'undefined'
- Renamed a variable early on in the code, and didn't update the later reference.

What part of the assignment did you spend the most time on?
Edge cases - e.g. what happens if there is more than one winner for the round.

Comfort Level (1-5):
4

Completeness Level (1-5):
4

What did you think of this deliverable?
OK

Is there anything in this code that you feel pleased about?
Was able to work around the array issues using isNaN()

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -51,6 +51,17 @@
...
+      <p>
+        First select the win condition: whether the largest or the smallest
+        number wins.
+      </p>
+      <p>Then input the number of players.</p>
+      <p>
+        The computer rolls the two dice for each player, and automatically
+        generates the best 2-digit number for the selected win condition. Good
+        luck!
+      </p>
```
Comment (lee-kimmixq) [inline index.html:64]: Good manipulation of HTML to give user instructions!

File: script.js [hunk 2]
```diff
@@ -1,4 +1,156 @@
...
+    result = String(smallerN) + String(largerN);
+  }
+  return Number(result);
+};
+
+// assess winner(s) of the round
+var assessWinner = function (playersResultsArray) {
+  var winningNumber;
+  if (gameMode == "largest") {
+    winningNumber = Math.max.apply(Math, playersResultsArray);
```
Comment (lee-kimmixq) [inline script.js:43]: Good use of `.apply()` method!

File: script.js [hunk 3]
```diff
@@ -1,4 +1,156 @@
...
+  while (i < playersResultsArray.length) {
+    if (playersResultsArray[i] == winningNumber) {
+      if (isNaN(winsArray[i])) {
+        winsArray[i] = 1;
+      } else {
+        winsArray[i] += 1;
+      }
+    }
+    i += 1;
+  }
```
Comment (lee-kimmixq) [inline script.js:58]: Good application of while loop! (Since you already know how to use for loops, you can just standardise to using a for loop!)

File: script.js [hunk 4]
```diff
@@ -1,4 +1,156 @@
...
+      Math.max.apply(Math, playersResultsArray) +
+      "<br>Min:" +
+      Math.min.apply(Math, playersResultsArray)
+  );
+  // run the assessWinner function to update the winsArray global variable
+  assessWinner(playersResultsArray);
+
+  // return output message for all players
+  return output;
+};
```
Comment (lee-kimmixq) [inline script.js:103]: Awesome!! This is an extremely useful and well-thought helper function that abstracts logic for one whole round!

File: script.js [hunk 5]
```diff
@@ -1,4 +1,156 @@
...
+        winrates += "Player " + (i + 1) + " hasn't won yet.<br>";
+      } else {
+        winrates +=
+          "Player " +
+          (i + 1) +
+          " has won " +
+          winsArray[i] +
+          " times so far.<br>";
+      }
+    }
```
Comment (lee-kimmixq) [inline script.js:147]: Good use of for loop! 👍

File: script.js [hunk 6]
```diff
@@ -1,4 +1,156 @@
...
+
+    // track the no. of rounds played
+    roundsPlayed += 1;
+
+    return (
+      currRound + "<br>Total rounds played: " + roundsPlayed + "<br>" + winrates
```
Comment (lee-kimmixq) [inline script.js:153]: Suggestion - use template literals instead!

File: script.js [hunk 7]
```diff
@@ -1,4 +1,156 @@
...
+      return "Please input type of win condition: 'largest' or 'smallest'.";
+    }
+  }
+  // input validation
+  if (
+    // invalid no. of players, return error message
+    (gameMode == "largest" || gameMode == "smallest") &&
+    !(input >= 2 && input <= 5) // game is configured to allow between 2 to 5 players
+  ) {
+    return "Invalid number of players";
```
Comment (lee-kimmixq) [inline script.js:127]: Nice input validation!

File: script.js [hunk 8]
```diff
@@ -1,4 +1,156 @@
...
+// loop thru all players and records their 2-digit number into an array
+var loopThruPlayers = function (nPlayers) {
+  var playersResultsArray = []; // declare an empty array for the current round
+  var output = "";
+  var counter = 0;
+  while (counter < nPlayers) {
+    var roll1 = rollDice();
+    var roll2 = rollDice();
+    // auto-generate the 2-digit number
+    var playerResult = combineNumber(gameMode, roll1, roll2);
```
Comment (lee-kimmixq) [inline script.js:74]: Suggestion - you can also shorten this to `var playerResult = combineNumber(gameMode, rollDice(), rollDice())`, since naming it `roll1` and `roll2` has not much significance here, it's more important in the `combineNumber()` helper function at line 20 instead!

File: script.js [hunk 9]
```diff
@@ -1,4 +1,156 @@
...
+  // input validation
+  if (
+    // invalid no. of players, return error message
+    (gameMode == "largest" || gameMode == "smallest") &&
+    !(input >= 2 && input <= 5) // game is configured to allow between 2 to 5 players
+  ) {
+    return "Invalid number of players";
+  } else {
+    // valid no. of players, proceed to play a round
+    var nPlayers = input;
```
Comment (lee-kimmixq) [inline script.js:130]: Interesting that we can change number of players every round!

File: script.js [hunk 10]
```diff
@@ -1,4 +1,156 @@
+/* This is a game of Beat That!, a 
+Each player attempts to get the largest/smallest 2-digit number from the concatenation of two consecutive dice rolls
+This game has been set up to allow between 2 to 5 players */
+
+// global variables
+var gameMode = "setup"; // initialize game mode
+var winsArray = []; // array to store players' wins
```
Comment (lee-kimmixq) [inline script.js:7]: Regarding the first error, you mentioned in your survey, you can avoid that situation by initialising `winsArray` as `[0,0,0,0,0]`

**Review by lee-kimmixq — COMMENTED** [review]
Super impressive work Tim! 💯 
Your code and logic is clean and concise (with very helpful comments!), and you made good use of helper functions! I think you already have a very good understanding of the concepts learnt and are able to apply these concepts very well :)

Your code is already awesome, but I left some suggestions to further shorten/simplify your code. Some other things you can try to explore are:
1. Store rolls in an array and use `.sort()` method to get the largest/smallest number
2. Use [template literals](https://basics.rocketacademy.co/4-getting-started-with-code/4.2-our-first-program#output-formatting) to format your strings! It reduces the possibility of syntax error due to missing + signs or quotation marks (as well as makes your code shorter and more readable)

File: script.js (hunk 6 above)
Comment (timochow) [reply script.js:153]: thanks, yes it would make for easier-to-read code and less prone to errors in formatting (e.g. missing whitespace)
